bsp/src/org/jetbrains/bsp/protocol/session/CancellableFuture.scala

```diff
@@ -0,0 +1,44 @@
+package org.jetbrains.bsp.protocol.session
+
+import java.util.concurrent.CompletableFuture
+
+/**
+ * JDK's CompletableFutre does not handle cancellation well.
+ * When a future does have an overriden `cancel` method, it is lost
+ * when doing transformations like `thenApply` or `thenCompose`.
+ *
+ * In Java 9, the new CompletableFuture's method was introduced: `newIncompleteFuture`.
+ * When using it, the cancel method of the original future will be called event when the
+ * `cancel` is done on a transformed one.
+ *
+ * When we know about CompletableFutures that override `cancel` methods, we should convert them
+ * to CancellableFuture before calling methods like `thenCompose`
+ *
+ * @param original
+ */
+class CancellableFuture[T](original: CompletableFuture[_]) extends CompletableFuture[T] {
+  override def cancel(mayInterruptIfRunning: Boolean): Boolean = {
+    original.cancel(mayInterruptIfRunning)
+    super.cancel(mayInterruptIfRunning)
+  }
+
+  override def newIncompleteFuture[U](): CompletableFuture[U] = {
+    new CancellableFuture(original)
+  }
+}
+
+object CancellableFuture {
+  def from[T](original: CompletableFuture[T]): CancellableFuture[T] = {
+    val result = new CancellableFuture[T](original)
+    original.whenComplete{ (value, error) =>
+      if(value != null) {
+        result.complete(value)
+      }
+      else {
+        result.completeExceptionally(error)
+      }
```

Comment:
@marek1840 I'm not sure if this is correct. Maybe it should be like this?

      if(error != null) {
        result.completeExceptionally(error)
      }
      else {
        result.complete(value)
      }

Reply:
Ok, changed this, as null exception cannot be thrown https://stackoverflow.com/questions/17576922/why-can-i-throw-null-in-java